File: cpp/tests/test_bootstrap.cpp
// Tests for the sequential bootstrap: round-trip repricing of every input
// instrument, schedule generation, validation and arbitrage detection.

#include "irm/bootstrap.hpp"

#include <cmath>
#include <limits>
#include <stdexcept>
#include <string>
#include <variant>
#include <vector>

#include <gtest/gtest.h>

#include "irm/quotes.hpp"

namespace {

const std::string kDataDir = IRM_DATA_DIR;

TEST(AnnualSchedule, KnownCases) {
    EXPECT_EQ(irm::annual_schedule(2.0), (std::vector<double>{1.0, 2.0}));
    EXPECT_EQ(irm::annual_schedule(0.25), (std::vector<double>{0.25}));
    const auto stub = irm::annual_schedule(2.5);
    ASSERT_EQ(stub.size(), 3u);
    EXPECT_NEAR(stub[0], 0.5, 1e-15);
    EXPECT_NEAR(stub[1], 1.5, 1e-15);
    EXPECT_NEAR(stub[2], 2.5, 1e-15);
    EXPECT_THROW(irm::annual_schedule(0.0), std::invalid_argument);
    EXPECT_THROW(irm::annual_schedule(-1.0), std::invalid_argument);
}

TEST(Instruments, ConstructionValidation) {
    EXPECT_THROW(irm::Deposit(0.0, 0.05), std::invalid_argument);
    EXPECT_THROW(irm::Deposit(1.0, std::nan("")), std::invalid_argument);
    // 1 + R*T <= 0 implies a non-positive DF.
    EXPECT_THROW(irm::Deposit(2.0, -0.5), std::invalid_argument);
    EXPECT_THROW(irm::FRA(-0.5, 1.0, 0.02), std::invalid_argument);
    EXPECT_THROW(irm::FRA(1.0, 1.0, 0.02), std::invalid_argument);
    EXPECT_THROW(irm::FRA(1.0, 2.0, -1.5), std::invalid_argument);
    EXPECT_THROW(irm::Swap(-2.0, 0.02), std::invalid_argument);
    EXPECT_THROW(irm::OISSwap(2.0, std::nan("")), std::invalid_argument);
}

TEST(Bootstrap, EmptyAndOrderingViolationsThrow) {
    EXPECT_THROW(irm::bootstrap({}), std::invalid_argument);
    // Duplicate pillar.
    EXPECT_THROW(irm::bootstrap({irm::Deposit(1.0, 0.02), irm::Deposit(1.0, 0.021)}),
                 std::invalid_argument);
    // Out of order.
    EXPECT_THROW(irm::bootstrap({irm::Deposit(2.0, 0.02), irm::Deposit(1.0, 0.02)}),
                 std::invalid_argument);
}

TEST(Bootstrap, SingleDepositExactDf) {
    const auto curve = irm::bootstrap({irm::Deposit(1.0, 0.05)});
    EXPECT_NEAR(curve.df(1.0), 1.0 / 1.05, 1e-14);
}

TEST(Bootstrap, EurCurveRoundTripAndNegativeShortEnd) {
    const auto quotes = irm::load_curve_quotes(kDataDir + "/curve_quotes.csv", "EUR");
    const auto curve = irm::bootstrap(quotes);
    // Round trip: every instrument reprices to < 1e-9 (reference < 1e-15).
    for (const auto& ins : quotes) {
        EXPECT_LT(std::fabs(irm::instrument_residual(ins, curve)), 1e-9)
            << "pillar " << irm::instrument_pillar(ins);
    }
    // Negative rates at the short end: DF > 1 (assert supported!).
    EXPECT_GT(curve.df(1.0), 1.0);
    EXPECT_LT(curve.zero_rate(0.5), 0.0);
}

TEST(Bootstrap, UsdCurveRoundTripAndParConsistency) {
    const auto quotes = irm::load_curve_quotes(kDataDir + "/curve_quotes.csv", "USD");
    const auto curve = irm::bootstrap(quotes);
    for (const auto& ins : quotes) {
        EXPECT_LT(std::fabs(irm::instrument_residual(ins, curve)), 1e-9);
    }
    // The par-swap helper must reproduce the quoted 10y par rate.
    std::vector<double> times = {0.0};
    for (int i = 1; i <= 10; ++i) times.push_back(static_cast<double>(i));
    EXPECT_NEAR(irm::par_swap_rate(curve, times), 0.0425, 1e-10);
    // Positive-rate curve: DFs strictly decreasing across pillars.
    double prev = 1.0;
    for (double t : curve.times()) {
        EXPECT_LT(curve.df(t), prev);
        prev = curve.df(t);
    }
}

TEST(Bootstrap, OisCurveRoundTrip) {
    const auto quotes = irm::load_ois_quotes(kDataDir + "/ois_quotes.csv");
    const auto curve = irm::bootstrap(quotes);
    for (const auto& ins : quotes) {
        EXPECT_LT(std::fabs(irm::instrument_residual(ins, curve)), 1e-9);
    }
}

TEST(Bootstrap, GapPillarsHandledByInterpolation) {
    // Swap-only long end with missing intermediate pillars: the 5y swap's
    // 1..4y coupons interpolate on the trial curve.
    const auto curve = irm::bootstrap({irm::Deposit(1.0, 0.03), irm::Swap(5.0, 0.035)});
    const irm::Swap five(5.0, 0.035);
    EXPECT_LT(std::fabs(five.residual(curve)), 1e-9);
    ASSERT_EQ(curve.times().size(), 2u);
}

TEST(Bootstrap, CrossedQuotesDetected) {
    // A 300% 2y par rate cannot be repriced by any positive DF(2): the
    // implied DF is negative -> domain_error naming the pillar.
    try {
        irm::bootstrap({irm::Deposit(1.0, 0.05), irm::Swap(2.0, 3.0)});
        FAIL() << "expected domain_error";
    } catch (const std::domain_error& exc) {
        EXPECT_NE(std::string(exc.what()).find("t=2.0"), std::string::npos);
        EXPECT_NE(std::string(exc.what()).find("crossed"), std::string::npos);
    }
}

TEST(Bootstrap, NonConvergenceMessageNotCrossedQuotes) {
    // A solver failure that is not a bracketing failure must not be
    // relabelled as crossed quotes (MINOR-10).
    try {
        irm::solve_pillar_df([](double x) { return (x > 1.0) ? INFINITY : -1.0; }, 2.0, "test");
        FAIL() << "expected domain_error";
    } catch (const std::domain_error& exc) {
        const std::string msg = exc.what();
        EXPECT_NE(msg.find("solver failed"), std::string::npos);
        EXPECT_NE(msg.find("not finite"), std::string::npos);
        EXPECT_EQ(msg.find("crossed"), std::string::npos);
    }
    try {
        irm::solve_pillar_df([](double x) { return 1.0 + x; }, 2.0, "test");
        FAIL() << "expected domain_error";
    } catch (const std::domain_error& exc) {
        EXPECT_NE(std::string(exc.what()).find("crossed"), std::string::npos);
    }
    EXPECT_NEAR(irm::solve_pillar_df([](double x) { return x - 0.9; }, 2.0, "test"), 0.9, 1e-13);
}

TEST(Instruments, MaturityOutOfRangeRejected) {
    // 1e12 used to be UB (float -> int) and a multi-GB allocation.
    for (double bad : {1e12, 1e9, irm::kMaxMaturity * (1.0 + 1e-9), 1e-13,
                       irm::kMinMaturity / 2.0, std::numeric_limits<double>::infinity()}) {
        EXPECT_THROW(irm::Swap(bad, 0.03), std::invalid_argument) << bad;
        EXPECT_THROW(irm::OISSwap(bad, 0.03), std::invalid_argument) << bad;
        EXPECT_THROW(irm::Deposit(bad, 0.03), std::invalid_argument) << bad;
        EXPECT_THROW(irm::FRA(0.0, bad, 0.03), std::invalid_argument) << bad;
        EXPECT_THROW(irm::annual_schedule(bad), std::invalid_argument) << bad;
    }
    EXPECT_EQ(irm::annual_schedule(irm::kMaxMaturity).size(), 200u);
    EXPECT_EQ(irm::annual_schedule(irm::kMinMaturity), (std::vector<double>{irm::kMinMaturity}));
    EXPECT_EQ(irm::annual_schedule(199.5).size(), 200u);
}

TEST(Bootstrap, FraStripReprices) {
    const std::vector<irm::Instrument> ins = {irm::Deposit(0.5, 0.030), irm::FRA(0.5, 1.0, 0.032),
                                              irm::FRA(1.0, 1.5, 0.034), irm::Swap(3.0, 0.033)};
    const auto curve = irm::bootstrap(ins);
    for (const auto& i : ins) EXPECT_LT(std::fabs(irm::instrument_residual(i, curve)), 1e-9);
    EXPECT_NEAR(curve.fwd_rate(0.5, 1.0), 0.032, 1e-12);
}

TEST(Bootstrap, OisSubYearMaturity) {
    const auto c = irm::bootstrap({irm::OISSwap(1.0 / 12.0, 0.03)});
    EXPECT_NEAR(c.df(1.0 / 12.0), 1.0 / (1.0 + 0.03 / 12.0), 1e-12);
}

TEST(Bootstrap, Dv01LocalisesInQuoteSpace) {
    const auto quotes = irm::load_curve_quotes(kDataDir + "/curve_quotes.csv", "USD");
    std::vector<double> times = {0.0};
    for (int i = 1; i <= 10; ++i) times.push_back(static_cast<double>(i));
    const auto bumped = [&](double pillar) {
        std::vector<irm::Instrument> out;
        for (const auto& q : quotes) {
            const double bump = (irm::instrument_pillar(q) == pillar) ? 1e-4 : 0.0;
            if (std::holds_alternative<irm::Deposit>(q)) {
                const auto& d = std::get<irm::Deposit>(q);
                out.emplace_back(irm::Deposit(d.maturity, d.rate + bump));
            } else {
                const auto& s = std::get<irm::Swap>(q);
                out.emplace_back(irm::Swap(s.maturity, s.rate + bump));
            }
        }
        return irm::bootstrap(out);
    };
    EXPECT_NEAR(irm::par_swap_rate(irm::bootstrap(quotes), times), 0.0425, 1e-12);
    EXPECT_NEAR(irm::par_swap_rate(bumped(5.0), times), 0.0425, 1e-12);
    EXPECT_NEAR(irm::par_swap_rate(bumped(10.0), times) - 0.0425, 1e-4, 1e-12);
}

TEST(Quotes, LoadersValidate) {
    EXPECT_THROW(irm::load_curve_quotes(kDataDir + "/curve_quotes.csv", "GBP"),
                 std::invalid_argument);
    EXPECT_THROW(irm::load_curve_quotes(kDataDir + "/does_not_exist.csv", "EUR"),
                 std::invalid_argument);
    const auto ys = irm::load_zero_yields(kDataDir + "/zero_yields.csv");
    EXPECT_EQ(ys.first.size(), 10u);
    EXPECT_EQ(ys.first.size(), ys.second.size());
}

}  // namespace

File: python/tests/test_bootstrap.py
"""Bootstrap: round-trip repricing, edge cases, invalid-quote detection."""

from __future__ import annotations

import pytest

from irm import (
    FRA,
    MAX_MATURITY,
    MIN_MATURITY,
    Deposit,
    DiscountCurve,
    OISSwap,
    Swap,
    annual_schedule,
    bootstrap,
    load_curve_quotes,
    load_ois_quotes,
    par_swap_rate,
    solve_pillar_df,
)


def test_eur_round_trip(eur_curve: DiscountCurve, data_dir) -> None:
    instruments = load_curve_quotes(data_dir / "curve_quotes.csv", "EUR")
    for ins in instruments:
        assert abs(ins.residual(eur_curve)) < 1e-9
    assert len(eur_curve.times) == len(instruments)


def test_usd_round_trip(usd_curve: DiscountCurve, data_dir) -> None:
    instruments = load_curve_quotes(data_dir / "curve_quotes.csv", "USD")
    for ins in instruments:
        assert abs(ins.residual(usd_curve)) < 1e-9


def test_ois_round_trip(ois_curve: DiscountCurve, data_dir) -> None:
    for ins in load_ois_quotes(data_dir / "ois_quotes.csv"):
        assert abs(ins.residual(ois_curve)) < 1e-9


def test_eur_negative_short_end_df_above_one(eur_curve: DiscountCurve) -> None:
    # Negative deposit rates imply DF > 1 — must be supported, not clipped.
    assert eur_curve.df(0.5) > 1.0
    assert eur_curve.df(1.0) > 1.0
    assert eur_curve.zero_rate(1.0) == pytest.approx(-0.003, abs=5e-4)


def test_usd_df_positive_and_decreasing(usd_curve: DiscountCurve) -> None:
    prev = 1.0
    for t in usd_curve.times:
        df = usd_curve.df(t)
        assert 0.0 < df < prev
        prev = df


def test_bootstrapped_par_rates_match_quotes(usd_curve: DiscountCurve, data_dir) -> None:
    # Property-style: every quoted swap's par rate off the final curve
    # equals its input quote (repricing in rate space).
    for ins in load_curve_quotes(data_dir / "curve_quotes.csv", "USD"):
        if isinstance(ins, Swap):
            times = [0.0] + list(annual_schedule(ins.maturity))
            assert par_swap_rate(usd_curve, times) == pytest.approx(
                ins.rate, abs=1e-10
            )


def test_bootstrap_with_fra_strip() -> None:
    instruments = [
        Deposit(0.5, 0.030),
        FRA(0.5, 1.0, 0.032),
        FRA(1.0, 1.5, 0.034),
        Swap(3.0, 0.033),
    ]
    curve = bootstrap(instruments)
    for ins in instruments:
        assert abs(ins.residual(curve)) < 1e-9
    # FRA repricing in rate space too.
    assert curve.fwd_rate(0.5, 1.0) == pytest.approx(0.032, abs=1e-12)


def test_deposit_only_and_swap_only() -> None:
    dep_curve = bootstrap([Deposit(0.25, 0.02), Deposit(0.5, 0.021), Deposit(1.0, 0.022)])
    assert dep_curve.df(1.0) == pytest.approx(1.0 / 1.022, rel=1e-12)
    swp = [Swap(1.0, 0.02), Swap(2.0, 0.022), Swap(5.0, 0.025), Swap(10.0, 0.027)]
    swp_curve = bootstrap(swp)  # gap pillars (3,4,6..9) come from interpolation
    for ins in swp:
        assert abs(ins.residual(swp_curve)) < 1e-9


def test_fractional_swap_maturity_schedule() -> None:
    assert annual_schedule(2.0) == (1.0, 2.0)
    assert annual_schedule(2.5) == (0.5, 1.5, 2.5)
    assert annual_schedule(0.25) == (0.25,)
    with pytest.raises(ValueError):
        annual_schedule(0.0)


def test_ordering_violations_rejected() -> None:
    with pytest.raises(ValueError, match="strictly increasing"):
        bootstrap([Deposit(1.0, 0.02), Deposit(0.5, 0.02)])
    with pytest.raises(ValueError, match="strictly increasing"):
        bootstrap([Deposit(1.0, 0.02), Swap(1.0, 0.02)])  # duplicate pillar
    with pytest.raises(ValueError, match="at least one"):
        bootstrap([])


def test_invalid_quotes_rejected_at_construction() -> None:
    with pytest.raises(ValueError, match="non-positive discount factor"):
        Deposit(1.0, -1.5)  # 1 + R*T <= 0 -> negative implied DF
    with pytest.raises(ValueError, match="non-positive discount factor"):
        FRA(1.0, 2.0, -1.2)
    with pytest.raises(ValueError):
        FRA(2.0, 1.0, 0.02)  # start >= end
    with pytest.raises(ValueError):
        Deposit(1.0, float("nan"))
    with pytest.raises(ValueError):
        Swap(-1.0, 0.02)


def test_crossed_quotes_fail_bracketing() -> None:
    # A -99% 2y swap after a sane 1y deposit needs DF(2) ~ 199 — outside any
    # admissible discount factor: bootstrap must raise, not return garbage.
    with pytest.raises(ValueError, match="crossed"):
        bootstrap([Deposit(1.0, 0.05), Swap(2.0, -0.99)])


class _NanInstrument:
    """Duck-typed instrument whose residual is NaN inside the bracket."""

    rate = 0.02

    @property
    def pillar(self) -> float:
        return 1.0

    def residual(self, curve: DiscountCurve) -> float:
        df = curve.df(1.0)
        if 0.5 < df < 1.5:
            return float("nan")
        return df - 0.97


def test_bootstrap_non_convergence_message_not_crossed_quotes() -> None:
    # A solver failure that is not a bracketing failure must not be
    # relabelled as crossed quotes (MINOR-10).
    with pytest.raises(ValueError) as exc:
        bootstrap([_NanInstrument()])
    msg = str(exc.value)
    assert "solver failed" in msg and "not finite" in msg
    assert "crossed" not in msg
    # The per-pillar kernel exposes the same classification directly.
    with pytest.raises(ValueError, match="crossed"):
        solve_pillar_df(lambda x: 1.0 + x, 2.0, "test")
    with pytest.raises(ValueError, match="solver failed"):
        solve_pillar_df(lambda x: float("inf") if x > 1.0 else -1.0, 2.0, "test")
    assert solve_pillar_df(lambda x: x - 0.9, 2.0, "test") == pytest.approx(0.9, abs=1e-13)


def test_swap_maturity_out_of_range_rejected() -> None:
    for bad in (1e12, 1e9, MAX_MATURITY * (1 + 1e-9), 1e-13, MIN_MATURITY / 2.0, float("inf")):
        with pytest.raises(ValueError):
            Swap(bad, 0.03)
        with pytest.raises(ValueError):
            OISSwap(bad, 0.03)
        with pytest.raises(ValueError):
            Deposit(bad, 0.03)
        with pytest.raises(ValueError):
            FRA(0.0, bad, 0.03)
        with pytest.raises(ValueError):
            annual_schedule(bad)
    # The bounds themselves are accepted and produce sane schedules.
    assert len(annual_schedule(MAX_MATURITY)) == 200
    assert annual_schedule(MIN_MATURITY) == (MIN_MATURITY,)
    assert len(annual_schedule(199.5)) == 200


def test_dv01_localises_in_quote_space(data_dir) -> None:
    quotes = load_curve_quotes(data_dir / "curve_quotes.csv", "USD")
    times = [0.0] + [float(i) for i in range(1, 11)]

    def bumped(pillar: float, bump: float = 1e-4):
        return bootstrap([
            type(q)(maturity=q.maturity, rate=q.rate + (bump if q.pillar == pillar else 0.0))
            for q in quotes
        ])

    base = par_swap_rate(bootstrap(quotes), times)
    assert base == pytest.approx(0.0425, abs=1e-12)
    # Bumping the 5y quote leaves the 10y par rate untouched ...
    assert par_swap_rate(bumped(5.0), times) == pytest.approx(0.0425, abs=1e-12)
    # ... bumping the 10y quote moves it by exactly the bump.
    assert par_swap_rate(bumped(10.0), times) - 0.0425 == pytest.approx(1e-4, abs=1e-12)


def test_ois_short_stub_schedule() -> None:
    # A 1M OIS has a single fixed payment with accrual = maturity.
    c = bootstrap([OISSwap(1.0 / 12.0, 0.03)])
    t = 1.0 / 12.0
    assert c.df(t) == pytest.approx(1.0 / (1.0 + 0.03 * t), rel=1e-12)
